karoo_gp/util/lazy_loader.py

```diff
@@ -0,0 +1,44 @@
+# This module is copied wholesale from https://wil.yegelwel.com/lazily-importing-python-modules/,
+# which is copied wholesale from the tensorflow library: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/util/lazy_loader.py
```

Comment:
It seems a bit backwards that in order to import the lazy loader to speed up imports you need to go through the slow import in the first place...

One issue with including the file here is that we have to worry about license.  They seem to use Apache 2.0, which is quite flexible, but it does have some requirements (e.g. including a copy of the whole license): https://tldrlegal.com/license/apache-license-2.0-(apache-2.0)

I investigated a bit, and the long import times might be because of the lack of GPU, so it might be faster on machines that have it.  Maybe we can create a separate issue and come back to it later?  Another simpler solution might be to just move the TensorFlow engine in a separate file, and import that file only if and when the user selects the TF engine.  That import will still be slow, but it will keep Karoo fast when other engines are used.

Reply:
> Another simpler solution might be to just move the TensorFlow engine in a separate file, and import that file only if and when the user selects the TF engine. That import will still be slow, but it will keep Karoo fast when other engines are used.

I think I tried that, both were in a `engine/` directory, and it had an `__init__.py` file that I imported both engines. Still, if you accessed the `__init__.py` file to get the `NumpyEngine`, `TensorflowEngine`, and therefore `tf`, was loaded.

Can I do without the `__init__.py`?

Reply:
> Can I do without the `__init__.py`?

If you have an empty `__init__.py` (required in order to access the modules in the dir as part of the package) it won't import anything.  If you don't have `from . import TensorFlowEngine` in the `__init__.py` then you would have to do e.g. `from karoo_gp.engines.tensorflow import TensorFlowEngine` instead of just doing `from karoo_gp.engines import TensorFlowEngine`.

If it wasn't for this issue, I think a single `engines.py` module would be ok.  You could also import tf in the `__init__` method of the `TensorFlowEngine` and/or in the other methods -- not particularly elegant but it should allow you to import the `engines.py` module without importing tf.